### PR TITLE
Move System.Text.RegularExpressions tests to xunit

### PR DESCRIPTION
**Pull request by hughbe**
This commit moves the tests for System.Text.RegularExpressions to use
xunit, instead of Console.WriteLine and really obfuscated error messages.
I have not added or removed any tests. I did, however, remove some obsolete test code for branches that would never occur.

Already, this test suite looks nicer. There is more to be done, however:
If/when this is merged, I will submit another PR consolidating common test code, and further modernizations including:
- Renaming tests and test files to make sense
- Splitting test data and test code
- Consolidating tests into various helper methods.

This PR contributes to #516 and #915

### CODE REVIEW

File: src/System.Text.RegularExpressions/tests/CharacterClassSubtractionSimple.cs [hunk 1]
```diff
@@ -8,40 +8,13 @@
...
-            ///////////////////////////////////////////////////////////////////
-            /////////////////////////// END TESTS /////////////////////////////
-        }
-        catch (Exception exc_general)
+        for (int i = 0; i < s_regexTests.Length; i++)
         {
-            ++iCountErrors;
-            Console.WriteLine("Error Err_8888yyy!  strLoc==" + strLoc + ", exc_general==" + exc_general.ToString());
+            Assert.True(s_regexTests[i].Run());
         }
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/CharacterClassSubtractionSimple.cs]: Seems like this RegexTestClass thing can be unraveled in the future, but for now, should this be a theory? e.g.
```C#
[Theory]
[MemberData(nameof(CharacterClassSubtractionTestCases))]
public void CharacterClassSubtraction(RegexTestCase case)
{
    case.Run();
}
```
?

File: src/System.Text.RegularExpressions/tests/R2LGetGroupNamesMatchTests.cs [hunk 2]
```diff
@@ -9,92 +9,20 @@
...
-        public static Match Match(string input);
-            "David Bau"
-
-        public static Boolean IsMatch(string input);     //D+
-            "12321"
-
+        public static String[] GetGroupNames(); "(?<first_name>\\S+)\\s(?<last_name>\\S+)"
+        public static Match Match(string input); "David Bau"
+        public static Boolean IsMatch(string input); //D+, "12321"
     */
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/R2LGetGroupNamesMatchTests.cs]: This seems unnecessary

File: src/System.Text.RegularExpressions/tests/RegexComponentsSplitTests.cs [hunk 3]
```diff
@@ -47,37 +47,37 @@ public static void RegexComponentsSplit()
         /*********************************************************
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexComponentsSplitTests.cs]: Can we clean this up?  Maybe with the theory approach mentioned elsewhere?

File: src/System.Text.RegularExpressions/tests/RegexConstructorTests.cs [hunk 4]
```diff
@@ -9,166 +9,30 @@
...
-            ///////////////////////////////////////////////////////////////////
-            /////////////////////////// END TESTS /////////////////////////////
-        }
-        catch (Exception exc_general)
-        {
-            ++iCountErrors;
-            Console.WriteLine("Error Err_8888yyy!  strLoc==" + strLoc + ", exc_general==" + exc_general.ToString());
-        }
+        // Should not throw
+        new Regex("foo", RegexOptions.ECMAScript | RegexOptions.IgnoreCase | RegexOptions.Multiline | RegexOptions.CultureInvariant);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexConstructorTests.cs]: Are there any properties we can assert on, e.g. to verify that what was passed in to the ctor roundtrips?

File: src/System.Text.RegularExpressions/tests/RegexConstructorTests.cs [hunk 5]
```diff
@@ -9,166 +9,30 @@
...
 
-        ////  Finish Diagnostics
-        Assert.Equal(0, iCountErrors);
+    [Fact]
+    public static void Ctor_Invalid()
+    {
+        // Pattern is null
+        Assert.Throws<ArgumentNullException>(() => new Regex(null));
+        Assert.Throws<ArgumentNullException>(() => new Regex(null, RegexOptions.None));
+        Assert.Throws<ArgumentNullException>(() => new Regex(null, RegexOptions.None, new TimeSpan()));
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexConstructorTests.cs]: My standard question: are there argument names we can validate in all of these argument exceptions? :)

File: src/System.Text.RegularExpressions/tests/RegexLangElementsCoverageTests.cs [hunk 6]
```diff
@@ -13,47 +13,10 @@ public class RegexLangElementsCoverageTests
...
-            ///////////////////////////////////////////////////////////////////
-            /////////////////////////// END TESTS /////////////////////////////
-        }
-        catch (Exception exc_general)
+        for (int i = 0; i < s_regexTests.Length; i++)
         {
-            ++iCountErrors;
-            Console.WriteLine("Error Err_8888yyy!  strLoc==" + strLoc + ", exc_general==" + exc_general.ToString());
+            Assert.True(s_regexTests[i].Run());
         }
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexLangElementsCoverageTests.cs]: Same theory question.  I'll stop mentioning these.

File: src/System.Text.RegularExpressions/tests/RegexMatchTests0.cs [hunk 7]
```diff
@@ -9,183 +9,75 @@
...
-
-        public static Match Match(string input, string pattern);     Testing control character escapes???
-            "2", "(\u0032)"
-
+    public static Match Match(string input, string pattern);
+        - Testing \B special character escape: "adfadsfSUCCESSadsfadsf", ".*\\B(SUCCESS)\\B.*"
+        - Testing octal sequence matches: "011", "\\060(\\061)?\\061"
+        - Testing hexadecimal sequence matches: "012", "(\\x30\\x31\\x32)"
+        - Testing control character escapes???: "2", "(\u0032)"
     */
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexMatchTests0.cs]: I think we can remove these "Tested methods" comments across the board, unless you think they add value.

File: src/System.Text.RegularExpressions/tests/RegexMatchTests10.cs [hunk 8]
```diff
@@ -9,127 +9,57 @@
...
-        public static Match Match(string input); Trying empty string. : Actual    "[a-z0-9]+"
-            ""
-
+        - Searching for numeric characters : Actual "[0-9]", "12345asdfasdfasdfljkhsda67890"
+        - Different pattern specification. This time range of symbols is allowed. : Actual "[a-z0-9]+", "[token1]? GARBAGEtoken2GARBAGE ;token3!"
+        - Trying empty string: Actual "[a-z0-9]+", ""
     */
-
     [Fact]
     public static void MatchTestCase10()
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexMatchTests10.cs]: Is there something more specific we can use for the test method name?

File: src/System.Text.RegularExpressions/tests/RegexMatchTests11.cs [hunk 9]
```diff
@@ -9,219 +9,84 @@
...
-            "209.25.0.111"
-
+    public Match Match(string input);
+        - Numbering pattern slots: "(?< 1>\\d{3})(?< 2>\\d{3})(?< 3>\\d{4})", "8885551111"
+        - Not naming pattern slots at all: "^(cat|chat)", "cats are bad"
+        - "([0-9]+(\\.[0-9]+){3})", "209.25.0.111"
     */
-
     [Fact]
     public static void RegexMatchTestCase11()
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexMatchTests11.cs]: Same naming question... I'll stop commenting on these... would seem to apply to all of the RegexMatchTests#.cs files

Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexMatchTests11.cs]: Details from the comments I've suggested deleting could be included in the test name.

---
In reply to: [56724437](https://github.com/dotnet/corefx/pull/6515#discussion_r56724437) [](ancestors = 56724437)

File: src/System.Text.RegularExpressions/tests/RegexMatchTests4.cs [hunk 10]
```diff
@@ -9,337 +9,129 @@
...
-        public  static  Match  Match ( string  input ) ; Nonbacktracking  subexpression :  Actual - " [ ^ 0 - 9 ] + ( ?>[0-9]+)3"
-            "abc123"
-
+    public Match Match(string input);
+        - Noncapturing group: Actual - "(a+)(?:b*)(ccc)", "aaabbbccc"
+        - Zero-width positive lookahead assertion: Actual - "abc(?=XXX)\\w+", "abcXXXdef"
+        - Zero-width negative lookahead assertion: Actual - "abc(?!XXX)\\w+", "abcXXXdef" - Negative
+        - Zero-width positive lookbehind assertion: Actual - "(\\w){6}(?<  =  XXX ) def ", " abcXXXdef "
+        - Zero-width  negative  lookbehind  assertion: Actual - " ( \ \ w ) { 6 } ( ? < ! XXX ) def ", " XXXabcdef "
+        - Nonbacktracking subexpression: Actual - " [ ^ 0 - 9 ] + ( ?>[0-9]+)3", "abc123"
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexMatchTests4.cs]: Should we split the single test method below into one test for each of these, named appropriately?

**Comment by stephentoub** [comment]
Thanks, @hughbe. What you've done here is a definite improvement over what was there before.  I do think there's more cleanup to be done; up to you whether you want to do that as part of this PR, or leave it for you / someone else for a future time.

File: src/System.Text.RegularExpressions/tests/RegexMatchTests4.cs (hunk 10 above)
Comment (hughbe) [inline src/System.Text.RegularExpressions/tests/RegexMatchTests4.cs]: Can we leave as-is for now, as this will be completely overhauled in a future PR?

**Comment by hughbe** [comment]
Yeah, I've got a lot more cleanup to do here!
 I've already done it, and just waiting on merging this one before submitting that

**Pull request merged**